packages/server-wallet/src/db/migrations/20200814010138_app_bytecode.ts

```diff
@@ -0,0 +1,17 @@
+import * as Knex from 'knex';
+
+const appBytecode = 'app_bytecode';
+
+export async function up(knex: Knex): Promise<any> {
+  await knex.schema.createTable(appBytecode, function(table) {
+    table.string('chain_id').notNullable();
+
+    table.string('app_definition').notNullable();
+    table.text('app_bytecode').notNullable();
+    table.primary(['chain_id', 'app_definition']);
```

Comment:
In my experience, I have only used an auto-incrementing integer, provided by postgres, for the primary key. I'm not 100% sure about mixing primary key strategies. What's the benefit given here? Does it automatically create a uniqueness constraint on `['chain_id', 'app_definition']`?

Not saying we shouldn't do this, it's just something I'm unfamiliar with.

Reply:
> In my experience, I have only used an auto-incrementing integer, provided by postgres, for the primary key. I'm not 100% sure about mixing primary key strategies. What's the benefit given here? Does it automatically create a uniqueness constraint on `['chain_id', 'app_definition']`?
> 
> Not saying we shouldn't do this, it's just something I'm unfamiliar with.

Yup, it automatically creates a uniqueness constraint on `['chain_id', 'app_definition']` . I think this is the "right" way as the combination of `chain_id` and `app_definition` should always be a unique bytecode entry.

We could use the auto-incrementing integer for the primary key but I think the only benefit is that it's the same thing other tables do.